### PR TITLE
server, tikv: enable unit test run parallel and add parallel flag for go coverage test

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
make all go unit test support `-check.p true` flag so we can run go test more quickly

### What is changed and how it works?
- change several test suite/test cases into `SerialSuites` to enable  `-check.p true` flag 
- optimize some test cases for store/tikv package
- add `-check.p true` for task `gotest` in Makefile to make it run faster to speed up pipeline `idc-jenkins-ci-tidb/check_dev_2`

results:
before changes:
```
$ make gotest
github.com/pingcap/tidb/planner/core	11.878s
github.com/pingcap/tidb/server	24.737s
github.com/pingcap/tidb/store/tikv	52.393s
Total Cost: 2'22''
```

after changes:
```
make gotest
github.com/pingcap/tidb/planner/core	5.140s
github.com/pingcap/tidb/server	14.331s
github.com/pingcap/tidb/store/tikv	32.132s
Total Cost: 1'35''
```

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

File: server/statistics_handler_test.go [hunk 1]
```diff
@@ -141,7 +156,8 @@ func (ds *testDumpStatsSuite) TestDumpStatsAPI(c *C) {
 }
 
 func (ds *testDumpStatsSuite) prepareData(c *C) {
-	db, err := sql.Open("mysql", getDSN())
+
```
Comment (wjhuang2016) [inline server/statistics_handler_test.go]: Remove this line?

File: server/statistics_handler_test.go [hunk 2]
```diff
@@ -34,13 +34,18 @@ import (
...
+	store      kv.Storage
+	domain     *domain.Domain
+	port       uint
+	statusPort uint
 }
 
-var _ = Suite(new(testDumpStatsSuite))
+var _ = Suite(&testDumpStatsSuite{
+	port:       4012,
+	statusPort: 10099,
```
Comment (Deardrops) [inline server/statistics_handler_test.go]: I think special a port here is not a good idea.
We have a mounts of test case in `server/http_handle_test.go`, which use port `4001`.
If we want to make all of them run parallel, specific port for them by `4013` , `4014`, and so on is a wrong way.
Suggestion: use `net.Listen("127.0.0.1:0")` to bind a non-used port. Or assign a rand port( before use it, check if it is used by `net.Dial()`).

**Comment by glorv** [comment]
@Deardrops  I update some code at server package. By replace the user defined config with auto-generated port&status port, we will make sure there won't exist any config-collision anymore. And newer test cases can depends on this easily. PTAL again, thanks~

File: server/tidb_test.go [hunk 3]
```diff
@@ -296,9 +299,10 @@ func registerTLSConfig(configName string, caCertPath string, clientCertPath stri
 }
 
 func (ts *TidbTestSuite) TestSystemTimeZone(c *C) {
+	portConf := newTestServerClient()
 	tk := testkit.NewTestKit(c, ts.store)
 	cfg := config.NewConfig()
-	cfg.Port = 4002
+	cfg.Port = portConf.port
```
Comment (Deardrops) [inline server/tidb_test.go]:
```suggestion
	cfg.Port = newTestServerClient().port
```

File: server/server_test.go [hunk 4]
```diff
@@ -36,28 +38,67 @@ import (
...
+func (cli *testServerClient) postStatus(path, contentType string, body io.Reader) (*http.Response, error) {
+	reqURL := fmt.Sprintf("http://localhost:%d%s", cli.statusPort, path)
+	return http.Post(reqURL, contentType, body)
+}
+
 // getDSN generates a DSN string for MySQL connection.
-func getDSN(overriders ...configOverrider) string {
+func (cli *testServerClient) getDSN(overriders ...configOverrider) string {
 	var config = defaultDSNConfig
+	config.Addr = fmt.Sprintf("127.0.0.1:%d", cli.port)
```
Comment (Deardrops) [inline server/server_test.go]: How about:
```suggestion
config := mysql.Config{
	User:   "root",
	Net:    "tcp",
	Addr:   fmt.Sprintf("127.0.0.1:%d", cli.port),
	DBName: "test",
	Strict: true,
}
```

**Comment by glorv** [comment]
@Deardrops  thank you for the patiently review. I resolved most of the comments but one. Currently due to our implement of [leak cheaker](https://github.com/pingcap/tidb/blob/master/util/testleak/leaktest.go#L67) use a single global map to track functions, we can't running multi test suite at the same time. So we cannot add `CustomParallelSuiteFlag` flag to tidb/server package which will make leak tests fail. It seems a good idea that we should update the leak cheaker to track leak functions by test suite scope and allow run multi suites in parallel. I think it not vary hard to implement it.

PTAL again, thanks~

File: infoschema/tables_test.go [hunk 5]
```diff
@@ -55,14 +55,18 @@ import (
 )
 
 var _ = Suite(&testTableSuite{})
-var _ = SerialSuites(&testClusterTableSuite{testTableSuite: &testTableSuite{}})
+var _ = SerialSuites(&testClusterTableSuite{testTableSuiteBase: &testTableSuiteBase{}})
 
 type testTableSuite struct {
+	testTableSuiteBase
```
Comment (Deardrops) [inline infoschema/tables_test.go]:
```suggestion
	*testTableSuiteBase
```

File: server/tidb_test.go [hunk 6]
```diff
@@ -41,18 +41,37 @@ import (
 	"github.com/pingcap/tidb/util/testkit"
 )
 
-type TidbTestSuite struct {
+type tiDBTestSuite struct {
+	tiDBTestSuiteBase
```
Comment (Deardrops) [inline server/tidb_test.go]:
```suggestion
	*TiDBTestSuiteBase
```

File: server/tidb_test.go [hunk 7]
```diff
@@ -41,18 +41,37 @@ import (
 	"github.com/pingcap/tidb/util/testkit"
 )
 
-type TidbTestSuite struct {
+type tiDBTestSuite struct {
```
Comment (Deardrops) [inline server/tidb_test.go]: These are `TiDB` or `tidb` only in our code naming, no `tiDB`

File: server/tidb_test.go [hunk 8]
```diff
@@ -41,18 +41,37 @@ import (
...
+	tiDBTestSuiteBase
+}
+
+type tiDBTestSerialSuite struct {
+	tiDBTestSuiteBase
```
Comment (Deardrops) [inline server/tidb_test.go]: ditto

**Comment by glorv** [comment]
@zz-jason  PTAL again. The latest 5 `run-all-tests` have executed successfully

**Pull request merged**